File: leetcode.com/0066_plus_one/solution.go
package main

import "fmt"

func plusOne(digits []int) []int {

    for i := len(digits) - 1 ; i >= 0 ; i-- {
        fmt.Println(i, digits[i])

        if digits[i] + 1 <= 9 {
            digits[i]++
            break
        }

        digits[i] = 0

        if i == 0 {
            digits = append([]int{1}, digits...)
        }
    }

    return digits
}

func main() {
    fmt.Println(plusOne([]int{9,9}))
}

File: leetcode.com/0066_plus_one/solution.c
int* plusOne(int* digits, int digitsSize, int* returnSize)
{
    bool extend = false;
    int out_size = digitsSize + 1;
    int* out_digits = (int*) calloc(sizeof(int), out_size);

    /* copy input */
    for (int i = digitsSize - 1 ; i >= 0 ; i--)
    {
        out_digits[i + 1] = digits[i];
    }

    /* increment */
    for (int i = out_size - 1 ; i >= 0 ; i--)
    {
        if (out_digits[i] + 1 <= 9) {
            out_digits[i]++;
            break;
        }

        out_digits[i] = 0;

        if (i == 1)
            extend = true;
    }

    if (extend == false) {
         for (int i = 0 ; i < out_size - 1 ; i++) {
             out_digits[i] = out_digits[i + 1];
         }

        out_size--;
    }

    *returnSize = out_size;
    return out_digits;
}
